### PR TITLE
reset end_of_dataloader for dataloader_dispatcher

### PR DESCRIPTION
**Pull request by megavaz**
Reset `end_of_dataloader` to `False` at each iteration as in [PR#1562](https://github.com/huggingface/accelerate/pull/1562) but for the `DataLoaderDispatcher`.

### CODE REVIEW

**Comment by megavaz** [comment]
Without the thing `lr_scheduler.step()` works incorrectly when gradient accumulation is used with `IterableDataset`

File: tests/test_data_loader.py [hunk 1]
```diff
@@ -384,3 +386,13 @@ def test_end_of_dataloader(self):
         # Test it also works on the second iteration
         for idx, _ in enumerate(dataloader):
             self.assertEqual(dataloader.end_of_dataloader, idx == 3)
+
+    def test_end_of_dataloader_dispatcher(self):
+        accelerator = Accelerator()
```
Comment (megavaz) [inline tests/test_data_loader.py:391]: Couldn't make the test work without the line. `AcceleratorState` raises "Please make sure to properly initialize your accelerator via `accelerator = Accelerator()` before using any functionality from the `accelerate` library."

**Review by sgugger — APPROVED** [review]
Thanks a lot for catching this!
